arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/result.kt

```diff
@@ -0,0 +1,68 @@
+package arrow.core.continuations
+
+import arrow.core.Option
+import arrow.core.identity
+import kotlin.contracts.ExperimentalContracts
+import kotlin.contracts.contract
+import kotlin.jvm.JvmInline
+
+public suspend fun <A> Effect<Throwable, A>.toResult(): Result<A> =
+  fold({ Result.failure(it) }, { Result.success(it) })
+
+public fun <A> EagerEffect<Throwable, A>.toResult(): Result<A> =
+  fold({ Result.failure(it) }, { Result.success(it) })
+
+@JvmInline
+public value class ResultEffectScope(private val cont: EffectScope<Throwable>) : EffectScope<Throwable> {
+  override suspend fun <B> shift(r: Throwable): B =
+    cont.shift(r)
+
+  public suspend fun <B> Option<B>.bind(): B =
+    bind { NullPointerException("Option is None") }
+
+  public suspend fun <B> Result<B>.bind(): B =
+    fold(::identity) { shift(it) }
+
+  public suspend fun ensure(value: Boolean): Unit =
+    ensure(value) { IllegalStateException("expected true, but was false") }
+}
+
+@JvmInline
+public value class ResultEagerEffectScope(private val cont: EagerEffectScope<Throwable>) : EagerEffectScope<Throwable> {
+  @Suppress("ILLEGAL_RESTRICTED_SUSPENDING_FUNCTION_CALL")
+  override suspend fun <B> shift(r: Throwable): B =
+    cont.shift(r)
+
+  public suspend fun <B> Option<B>.bind(): B =
+    bind { NullPointerException("Option is None") }
+
+  public suspend fun <B> Result<B>.bind(): B =
+    fold(::identity) { shift(it) }
+
+  public suspend fun ensure(value: Boolean): Unit =
+    ensure(value) { IllegalStateException("expected true, but was false") }
```

Comment:
@nomisRev is this Exception fine or should we a custom one, which doesn't include the stacktrace?
e.g.: `public class EnsureContinuationException(msg: String) : IllegalStateException(msg, null)` or something along the lines

Reply:
We could also omit these overrides and then we don't make a decision for the users, since they can use `ensure(...) { }` from `EffectScope<Throwable>`.

This is maybe a better initial approach until users request such behavior to be added.
Same for `Option<B>.bind()` and `ensureNotNull`.

cc\\ @pakoito

Reply:
Will add that instead